### PR TITLE
[WIP] Improve CI/CD pipeline performance

### LINKED ISSUE #24028: CI: Use all available cores in the CI
If I'm not wrong, both Travis and Azure provide two cores for the execution of jobs (I think in both cases the number does not necessarily need to be 2, and could be higher, or I guess lower, depending on their load).

So far we are just parallelizing the execution of tests, where we are using 2 processes in most cases (except for the tests marked as `single`).

We should be able to speed up the CI by detecting the number of cpus available for the job, and parallelize at least (assuming 2 cpus):
- Building pandas (adding `-j2` to `python setup.py build_ext --inplace`)
- Building the docs (adding `--num-jobs=2` to `./doc/make.py html`)
- Running the tests, with the detected cpus, and where it's not already using multiple processes
- Not sure if there is anything else that can be run in more than one core (conda, linting, benchmarks...)

Travis specs: https://docs.travis-ci.com/user/reference/overview/#virtualisation-environment-vs-operating-system

**UPDATE:**

- Based on tests in my local machine, doesn't seem like the `-j` argument in `setup.py build_ext` has any effect. I'd say it's ignored, and the process is always single core.

Time to run the tests `not slow and not network` in the CI:

1st run:

|        | -j 0     | -j auto |
|--------|----------|---------|
| azure  | 1443 sec | 804 sec |
| travis | 1362 sec | 959 sec |

2nd run:

|        | -j 0     | -j auto |
|--------|----------|---------|
| azure  | 1461 sec | 839 sec |
| travis | 1421 sec | 1009 sec |

### PR DESCRIPTION
**Pull request by bhavaniravi**
- [X] closes #24028 
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Out of curiosity, how big is the perf difference?

**Comment by bhavaniravi** [comment]
> Out of curiosity, how big is the perf difference?

It is WIP. WIll know once I make some progress.

File: doc/source/whatsnew/v0.9.1.rst [hunk 1]
```diff
@@ -59,7 +59,7 @@ New features
 
         .. ipython:: python
 
-    	    df = DataFrame(np.random.randn(5, 3), columns = ['A','B','C'])
+    	    df = pd.DataFrame(np.random.randn(5, 3), columns = ['A','B','C'])
```
Comment (WillAyd) [inline doc/source/whatsnew/v0.9.1.rst]: Can you revert changes to this file?

**Review by WillAyd — COMMENTED** [review]
Seems logical but does the latest update actually change anything? Quick eye test seems similar to the rest of CI

**Comment by jreback** [comment]
does this change perf? can you merge master as well

File: doc/source/whatsnew/v0.9.1.rst (hunk 1 above)
Comment (bhavaniravi) [reply doc/source/whatsnew/v0.9.1.rst]: The doc build fails due to an import error

**Comment by bhavaniravi** [comment]
@WillAyd @jreback yes, it is not reflecting on the performance improvement. I was off a couple of weeks now I'm back on. Will check it

**Comment by jreback** [comment]
@bhavaniravi if you want to continue on this, pls merge master and post a comparison of before / after.

**Comment by WillAyd** [comment]
I think this is stale or not really giving the results we are after, so closing. @bhavaniravi if you disagree ping and can reopen for sure

Thanks as always!

**Pull request closed without merging**